src/cephadm/cephadm.py

```diff
@@ -3230,15 +3236,30 @@ def fetch_custom_config_files(ctx: CephadmContext) -> List[Dict[str, Any]]:
     return []
 
 
-def fetch_tcp_ports(ctx: CephadmContext) -> List[int]:
+def fetch_tcp_ports(ctx: CephadmContext) -> List[EndPoint]:
     """Return a list of tcp ports, as integers, stored on the given ctx.
     """
     ports = getattr(ctx, 'tcp_ports', None)
     if ports is None:
-        return []
+        ports = []
     if isinstance(ports, str):
-        return list(map(int, ports.split()))
-    return ports
+        ports = list(map(int, ports.split()))
+    port_ips: Dict[str, str] = {}
+    port_ips_attr: Optional[Union[str, Dict[str, str]]] = getattr(ctx, 'port_ips', None)
```

Comment:
Since you are already using a union you can simply the type as: `Union[str, Dict[str, str], None]`

Reply:
done